<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Importing from to tRNAdb and mitotRNAdb as `GRanges` • tRNAdbImport</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Importing from to tRNAdb and mitotRNAdb as `GRanges`">
<meta property="og:description" content="tRNAdbImport">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tRNAdbImport</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.21.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/tRNAdbImport.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/FelixErnst/tRNAdbImport/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Importing from to tRNAdb and mitotRNAdb as
<code>GRanges</code>
</h1>
                        <h4 data-toc-skip class="author">Felix G.M.
Ernst</h4>
            
            <h4 data-toc-skip class="date">2024-03-23</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/FelixErnst/tRNAdbImport/blob/HEAD/vignettes/tRNAdbImport.Rmd" class="external-link"><code>vignettes/tRNAdbImport.Rmd</code></a></small>
      <div class="hidden name"><code>tRNAdbImport.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The tRNAdb and mttRNAdb <span class="citation">(Jühling et al.
2009)</span> is a compilation of tRNA sequences and tRNA genes. It is a
follow up version of the database of Sprinzl et al.  <span class="citation">(Sprinzl and Vassilenko 2005)</span>.</p>
<p>Using <code>tRNAdbImport</code> the tRNAdb can be accessed as
outlined on the website <a href="http://trna.bioinf.uni-leipzig.de/" class="external-link">http://trna.bioinf.uni-leipzig.de/</a>
and the results are returned as a <code>GRanges</code> object.</p>
<div class="section level3">
<h3 id="status-2024">Status 2024<a class="anchor" aria-label="anchor" href="#status-2024"></a>
</h3>
<p>The tRNAdb Server is currently not available. Some chunks of the code
in this vignette are currently not avilable. See <a href="https://www.bioinf.uni-leipzig.de/services/webservices" class="external-link">https://www.bioinf.uni-leipzig.de/services/webservices</a>
for more information.</p>
</div>
</div>
<div class="section level2">
<h2 id="importing-as-granges">Importing as <code>GRanges</code><a class="anchor" aria-label="anchor" href="#importing-as-granges"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">tRNAdbImport</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># accessing tRNAdb</span></span>
<span><span class="co"># tRNA from yeast for Alanine and Phenylalanine</span></span>
<span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import.tRNAdb.html">import.tRNAdb</a></span><span class="op">(</span>organism <span class="op">=</span> <span class="st">"Saccharomyces cerevisiae"</span>,</span>
<span>                    aminoacids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Phe"</span>,<span class="st">"Ala"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: tRNAdb Server seems to be not available.</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get a Phenylalanine tRNA from yeast</span></span>
<span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import.tRNAdb.html">import.tRNAdb.id</a></span><span class="op">(</span>tdbID <span class="op">=</span> <span class="va">gr</span><span class="op">[</span><span class="va">gr</span><span class="op">$</span><span class="va">tRNA_type</span> <span class="op">==</span> <span class="st">"Phe"</span>,<span class="op">]</span><span class="op">[</span><span class="fl">1L</span><span class="op">]</span><span class="op">$</span><span class="va">tRNAdb_ID</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># find the same tRNA via blast</span></span>
<span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import.tRNAdb.html">import.tRNAdb.blast</a></span><span class="op">(</span>blastSeq <span class="op">=</span> <span class="va">gr</span><span class="op">$</span><span class="va">tRNA_seq</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># accessing mtRNAdb</span></span>
<span><span class="co"># get the mitochrondrial tRNA for Alanine in Bos taurus</span></span>
<span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import.tRNAdb.html">import.mttRNAdb</a></span><span class="op">(</span>organism <span class="op">=</span> <span class="st">"Bos taurus"</span>, </span>
<span>                      aminoacids <span class="op">=</span> <span class="st">"Ala"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: tRNAdb Server seems to be not available.</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get one mitochrondrial tRNA in Bos taurus. </span></span>
<span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import.tRNAdb.html">import.mttRNAdb.id</a></span><span class="op">(</span>mtdbID <span class="op">=</span> <span class="va">gr</span><span class="op">[</span><span class="fl">1L</span><span class="op">]</span><span class="op">$</span><span class="va">tRNAdb_ID</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># check that the result has the appropriate columns</span></span>
<span><span class="fu"><a href="../reference/istRNAdbGRanges.html">istRNAdbGRanges</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Input GRanges object does not meet the requirements of the function. The following columns are expected:</span></span>
<span><span class="co">## 'tRNA_length', 'tRNA_type', 'tRNA_anticodon', 'tRNA_seq', 'tRNA_str', 'tRNA_CCA.end', 'tRNAdb_ID', 'tRNAdb', 'tRNAdb_organism', 'tRNAdb_strain', 'tRNAdb_taxonomyID', 'tRNAdb_verified'.</span></span></code></pre>
<pre><code><span><span class="co">## [1] FALSE</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="importing-as-granges-from-the-rna-database">Importing as <code>GRanges</code> from the RNA database<a class="anchor" aria-label="anchor" href="#importing-as-granges-from-the-rna-database"></a>
</h2>
<p>The tRNAdb offers two different sets of data, one containing DNA
sequences and one containing RNA sequences. Depending on the database
selected, <code>DNA</code> as default, the GRanges will contain a
<code>DNAStringSet</code> or a <code>ModRNAStringSet</code> as the
<code>tRNA_seq</code> column. Because the RNA sequences can contain
modified nucleotides, the <code>ModRNAStringSet</code> class is used
instead of the <code>RNAStringSet</code> class to store the sequences
correctly with all information intact.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import.tRNAdb.html">import.tRNAdb</a></span><span class="op">(</span>organism <span class="op">=</span> <span class="st">"Saccharomyces cerevisiae"</span>,</span>
<span>                    aminoacids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Phe"</span>,<span class="st">"Ala"</span><span class="op">)</span>,</span>
<span>                    database <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span></span>
<span><span class="va">gr</span><span class="op">$</span><span class="va">tRNA_seq</span></span></code></pre></div>
<p>The special characters in the sequence might no exactly match the
ones shown on the website, since they are sanitized internally to a
unified dictionary defined in the <code>Modstrings</code> package.
However, the type of modification encoded will remain the same (See the
<code>Modstrings</code> package for more details).</p>
<p>The information on the position and type of the modifications can
also be converted into a tabular format using the <code>separate</code>
function from the <code>Modstrings</code> package.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Modstrings/man/separate.html" class="external-link">separate</a></span><span class="op">(</span><span class="va">gr</span><span class="op">$</span><span class="va">tRNA_seq</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="further-analysis">Further analysis<a class="anchor" aria-label="anchor" href="#further-analysis"></a>
</h2>
<p>The output can be saved or directly used for further analysis.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings" class="external-link">Biostrings</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span>
<span><span class="co"># saving the tRAN sequences as fasta file</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">writeXStringSet</a></span><span class="op">(</span><span class="va">gr</span><span class="op">$</span><span class="va">tRNA_seq</span>, filepath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># converting tRNAdb information to GFF compatible values</span></span>
<span><span class="va">gff</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import.tRNAdb.html">tRNAdb2GFF</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span>
<span><span class="va">gff</span></span>
<span><span class="co"># Saving the information as gff3 file</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html" class="external-link">export.gff3</a></span><span class="op">(</span><span class="va">gff</span>, con <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Please have a look at the <code>tRNA</code> package for further
analysis of the tRNA sequences.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R Under development (unstable) (2024-03-20 r86162)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] rtracklayer_1.63.1   tRNAdbImport_1.21.1  tRNA_1.21.1         </span></span>
<span><span class="co">##  [4] Structstrings_1.19.1 Modstrings_1.19.0    Biostrings_2.71.5   </span></span>
<span><span class="co">##  [7] XVector_0.43.1       GenomicRanges_1.55.4 GenomeInfoDb_1.39.9 </span></span>
<span><span class="co">## [10] IRanges_2.37.1       S4Vectors_0.41.5     BiocGenerics_0.49.1 </span></span>
<span><span class="co">## [13] BiocStyle_2.31.0    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] SummarizedExperiment_1.33.3 gtable_0.3.4               </span></span>
<span><span class="co">##  [3] rjson_0.2.21                xfun_0.42                  </span></span>
<span><span class="co">##  [5] bslib_0.6.2                 ggplot2_3.5.0              </span></span>
<span><span class="co">##  [7] httr2_1.0.0                 lattice_0.22-6             </span></span>
<span><span class="co">##  [9] Biobase_2.63.0              vctrs_0.6.5                </span></span>
<span><span class="co">## [11] tools_4.4.0                 bitops_1.0-7               </span></span>
<span><span class="co">## [13] generics_0.1.3              parallel_4.4.0             </span></span>
<span><span class="co">## [15] curl_5.2.1                  tibble_3.2.1               </span></span>
<span><span class="co">## [17] fansi_1.0.6                 pkgconfig_2.0.3            </span></span>
<span><span class="co">## [19] Matrix_1.7-0                desc_1.4.3                 </span></span>
<span><span class="co">## [21] lifecycle_1.0.4             GenomeInfoDbData_1.2.11    </span></span>
<span><span class="co">## [23] compiler_4.4.0              stringr_1.5.1              </span></span>
<span><span class="co">## [25] Rsamtools_2.19.4            textshaping_0.3.7          </span></span>
<span><span class="co">## [27] munsell_0.5.0               codetools_0.2-19           </span></span>
<span><span class="co">## [29] htmltools_0.5.7             sass_0.4.9                 </span></span>
<span><span class="co">## [31] RCurl_1.98-1.14             yaml_2.3.8                 </span></span>
<span><span class="co">## [33] pkgdown_2.0.7               pillar_1.9.0               </span></span>
<span><span class="co">## [35] crayon_1.5.2                jquerylib_0.1.4            </span></span>
<span><span class="co">## [37] BiocParallel_1.37.1         DelayedArray_0.29.9        </span></span>
<span><span class="co">## [39] cachem_1.0.8                abind_1.4-5                </span></span>
<span><span class="co">## [41] tidyselect_1.2.1            digest_0.6.35              </span></span>
<span><span class="co">## [43] stringi_1.8.3               restfulr_0.0.15            </span></span>
<span><span class="co">## [45] dplyr_1.1.4                 purrr_1.0.2                </span></span>
<span><span class="co">## [47] bookdown_0.38               fastmap_1.1.1              </span></span>
<span><span class="co">## [49] grid_4.4.0                  SparseArray_1.3.4          </span></span>
<span><span class="co">## [51] colorspace_2.1-0            cli_3.6.2                  </span></span>
<span><span class="co">## [53] magrittr_2.0.3              S4Arrays_1.3.6             </span></span>
<span><span class="co">## [55] XML_3.99-0.16.1             utf8_1.2.4                 </span></span>
<span><span class="co">## [57] scales_1.3.0                rappdirs_0.3.3             </span></span>
<span><span class="co">## [59] httr_1.4.7                  rmarkdown_2.26             </span></span>
<span><span class="co">## [61] matrixStats_1.2.0           ragg_1.3.0                 </span></span>
<span><span class="co">## [63] memoise_2.0.1               evaluate_0.23              </span></span>
<span><span class="co">## [65] knitr_1.45                  BiocIO_1.13.0              </span></span>
<span><span class="co">## [67] rlang_1.1.3                 glue_1.7.0                 </span></span>
<span><span class="co">## [69] BiocManager_1.30.22         xml2_1.3.6                 </span></span>
<span><span class="co">## [71] jsonlite_1.8.8              R6_2.5.1                   </span></span>
<span><span class="co">## [73] MatrixGenerics_1.15.0       GenomicAlignments_1.39.4   </span></span>
<span><span class="co">## [75] systemfonts_1.0.6           fs_1.6.3                   </span></span>
<span><span class="co">## [77] zlibbioc_1.49.3</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Juehling.2009" class="csl-entry">
Jühling, Frank, Mario Mörl, Roland K. Hartmann, Mathias Sprinzl, Peter
F. Stadler, and Joern Pütz. 2009. <span>“tRNAdb 2009: Compilation of
tRNA Sequences and tRNA Genes.”</span> <em>Nucleic Acids Research</em>
37 (suppl_1): D159–62. <a href="https://doi.org/10.1093/nar/gkn772" class="external-link">https://doi.org/10.1093/nar/gkn772</a>.
</div>
<div id="ref-Sprinzl.2005" class="csl-entry">
Sprinzl, Mathias, and Konstantin S. Vassilenko. 2005. <span>“Compilation
of tRNA Sequences and Sequences of tRNA Genes.”</span> <em>Nucleic Acids
Research</em> 33 (suppl_1): D139–40. <a href="https://doi.org/10.1093/nar/gki012" class="external-link">https://doi.org/10.1093/nar/gki012</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Felix G.M. Ernst.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
