<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Importing from to tRNAdb and mitotRNAdb as `GRanges` • tRNAdbImport</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Importing from to tRNAdb and mitotRNAdb as `GRanges`">
<meta property="og:description" content="tRNAdbImport">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tRNAdbImport</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.7.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/tRNAdbImport.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/FelixErnst/tRNAdbImport/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="tRNAdbImport_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Importing from to tRNAdb and mitotRNAdb as <code>GRanges</code>
</h1>
                        <h4 class="author">Felix G.M. Ernst</h4>
            
            <h4 class="date">2020-07-18</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/FelixErnst/tRNAdbImport/blob/master/vignettes/tRNAdbImport.Rmd"><code>vignettes/tRNAdbImport.Rmd</code></a></small>
      <div class="hidden name"><code>tRNAdbImport.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>The tRNAdb and mttRNAdb <span class="citation">(Jühling et al. 2009)</span> is a compilation of tRNA sequences and tRNA genes. It is a follow up version of the database of Sprinzl et al.  <span class="citation">(Sprinzl and Vassilenko 2005)</span>.</p>
<p>Using <code>tRNAdbImport</code> the tRNAdb can be accessed as outlined on the website <a href="http://trna.bioinf.uni-leipzig.de/">http://trna.bioinf.uni-leipzig.de/</a> and the results are returned as a <code>GRanges</code> object.</p>
</div>
<div id="importing-as-granges" class="section level1">
<h1 class="hasAnchor">
<a href="#importing-as-granges" class="anchor"></a>Importing as <code>GRanges</code>
</h1>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">tRNAdbImport</span>)
</pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="co"># accessing tRNAdb</span>
<span class="co"># tRNA from yeast for Alanine and Phenylalanine</span>
<span class="kw">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import.tRNAdb.html">import.tRNAdb</a></span>(organism = <span class="st">"Saccharomyces cerevisiae"</span>,
                    aminoacids = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Phe"</span>,<span class="st">"Ala"</span>))
</pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="co"># get a Phenylalanine tRNA from yeast</span>
<span class="kw">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import.tRNAdb.html">import.tRNAdb.id</a></span>(tdbID = <span class="kw">gr</span>[<span class="kw">gr</span><span class="op">$</span><span class="kw">tRNA_type</span> <span class="op">==</span> <span class="st">"Phe"</span>,][<span class="fl">1L</span>]<span class="op">$</span><span class="kw">tRNAdb_ID</span>)
</pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="co"># find the same tRNA via blast</span>
<span class="kw">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import.tRNAdb.html">import.tRNAdb.blast</a></span>(blastSeq = <span class="kw">gr</span><span class="op">$</span><span class="kw">tRNA_seq</span>)
</pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="co"># accessing mtRNAdb</span>
<span class="co"># get the mitochrondrial tRNA for Alanine in Bos taurus</span>
<span class="kw">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import.tRNAdb.html">import.mttRNAdb</a></span>(organism = <span class="st">"Bos taurus"</span>, 
                      aminoacids = <span class="st">"Ala"</span>)
</pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="co"># get one mitochrondrial tRNA in Bos taurus. </span>
<span class="kw">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import.tRNAdb.html">import.mttRNAdb.id</a></span>(mtdbID = <span class="kw">gr</span>[<span class="fl">1L</span>]<span class="op">$</span><span class="kw">tRNAdb_ID</span>)
</pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="co"># check that the result has the appropriate columns</span>
<span class="fu"><a href="../reference/istRNAdbGRanges.html">istRNAdbGRanges</a></span>(<span class="kw">gr</span>)
</pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="importing-as-granges-from-the-rna-database" class="section level1">
<h1 class="hasAnchor">
<a href="#importing-as-granges-from-the-rna-database" class="anchor"></a>Importing as <code>GRanges</code> from the RNA database</h1>
<p>The tRNAdb offers two different sets of data, one containing DNA sequences and one containing RNA sequences. Depending on the database selected, <code>DNA</code> as default, the GRanges will contain a <code>DNAStringSet</code> or a <code>ModRNAStringSet</code> as the <code>tRNA_seq</code> column. Because the RNA sequences can contain modified nucleotides, the <code>ModRNAStringSet</code> class is used instead of the <code>RNAStringSet</code> class to store the sequences correctly with all information intact.</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="kw">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import.tRNAdb.html">import.tRNAdb</a></span>(organism = <span class="st">"Saccharomyces cerevisiae"</span>,
                    aminoacids = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Phe"</span>,<span class="st">"Ala"</span>),
                    database = <span class="st">"RNA"</span>)
<span class="kw">gr</span><span class="op">$</span><span class="kw">tRNA_seq</span>
</pre></div>
<pre><code>##   A ModRNAStringSet instance of length 3
##     width seq                                               names               
## [1]    76 GGGCGUGUKGCGUAGDCGGDAGC...TPCGAUUCCGGACUCGUCCACCA tdbR00000012
## [2]    76 GCGGAUUUALCUCAGDDGGGAGA...TPCG"UCCACAGAAUUCGCACCA tdbR00000083
## [3]    76 GCGGACUUALCUCAGDDGGGAGA...TPCG"UCCACAGAGUUCGCACCA tdbR00000084</code></pre>
<p>The special characters in the sequence might no exactly match the ones shown on the website, since they are sanitized internally to a unified dictionary defined in the <code>Modstrings</code> package. However, the type of modification encoded will remain the same (See the <code>Modstrings</code> package for more details).</p>
<p>The information on the position and type of the modifications can also be converted into a tabular format using the <code>separate</code> function from the <code>Modstrings</code> package.</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="fu">separate</span>(<span class="kw">gr</span><span class="op">$</span><span class="kw">tRNA_seq</span>)
</pre></div>
<pre><code>## GRanges object with 38 ranges and 1 metadata column:
##            seqnames    ranges strand |         mod
##               &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##    [1] tdbR00000012         9      + |         m1G
##    [2] tdbR00000012        16      + |           D
##    [3] tdbR00000012        20      + |           D
##    [4] tdbR00000012        26      + |       m2,2G
##    [5] tdbR00000012        34      + |           I
##    ...          ...       ...    ... .         ...
##   [34] tdbR00000084        46      + |         m7G
##   [35] tdbR00000084        49      + |        f5Cm
##   [36] tdbR00000084        54      + |         m5U
##   [37] tdbR00000084        55      + |           Y
##   [38] tdbR00000084        58      + |         m1A
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="further-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#further-analysis" class="anchor"></a>Further analysis</h1>
<p>The output can be saved or directly used for further analysis.</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">Biostrings</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">rtracklayer</span>)
<span class="co"># saving the tRAN sequences as fasta file</span>
<span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">writeXStringSet</a></span>(<span class="kw">gr</span><span class="op">$</span><span class="kw">tRNA_seq</span>, filepath = <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempfile</a></span>())
<span class="co"># converting tRNAdb information to GFF compatible values</span>
<span class="kw">gff</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import.tRNAdb.html">tRNAdb2GFF</a></span>(<span class="kw">gr</span>)
<span class="kw">gff</span>
</pre></div>
<pre><code>## GRanges object with 3 ranges and 20 metadata columns:
##           seqnames    ranges strand |   source     type     score     phase
##              &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;factor&gt; &lt;integer&gt; &lt;integer&gt;
##   [1] tdbR00000012      1-76      * |   tRNAdb     tRNA      &lt;NA&gt;      &lt;NA&gt;
##   [2] tdbR00000083      1-76      * |   tRNAdb     tRNA      &lt;NA&gt;      &lt;NA&gt;
##   [3] tdbR00000084      1-76      * |   tRNAdb     tRNA      &lt;NA&gt;      &lt;NA&gt;
##                 ID        no tRNA_length   tRNA_type tRNA_anticodon
##        &lt;character&gt; &lt;integer&gt;   &lt;integer&gt; &lt;character&gt;    &lt;character&gt;
##   [1] tdbR00000012         1          76         Ala            IGC
##   [2] tdbR00000083         2          76         Phe            #AA
##   [3] tdbR00000084         3          76         Phe            #AA
##                                                                           tRNA_seq
##                                                                        &lt;character&gt;
##   [1] GGGCGUGUKGCGUAGDCGGDAGCGCRCUCCCUUIGCOPGGGAGAGGDCUCCGGTPCGAUUCCGGACUCGUCCACCA
##   [2] GCGGAUUUALCUCAGDDGGGAGAGCRCCAGABU#AAYAP°UGGAG7UC°UGUGTPCG"UCCACAGAAUUCGCACCA
##   [3] GCGGACUUALCUCAGDDGGGAGAGCRCCAGABU#AAYAP°UGGAG7UC°UGUGTPCG"UCCACAGAGUUCGCACCA
##                                                                           tRNA_str
##                                                                        &lt;character&gt;
##   [1] &lt;&lt;&lt;&lt;&lt;.&lt;..&lt;&lt;&lt;&lt;........&gt;&gt;&gt;&gt;.&lt;&lt;&lt;&lt;&lt;.......&gt;&gt;&gt;&gt;&gt;.....&lt;&lt;&lt;&lt;&lt;.......&gt;&gt;&gt;&gt;&gt;&gt;.&gt;&gt;&gt;&gt;&gt;....
##   [2] &lt;&lt;&lt;&lt;&lt;&lt;&lt;..&lt;&lt;&lt;&lt;........&gt;&gt;&gt;&gt;.&lt;&lt;&lt;&lt;&lt;.......&gt;&gt;&gt;&gt;&gt;.....&lt;&lt;&lt;&lt;&lt;.......&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;....
##   [3] &lt;&lt;&lt;&lt;&lt;&lt;&lt;..&lt;&lt;&lt;&lt;........&gt;&gt;&gt;&gt;.&lt;&lt;&lt;&lt;&lt;.......&gt;&gt;&gt;&gt;&gt;.....&lt;&lt;&lt;&lt;&lt;.......&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;....
##       tRNA_CCA.end      tRNAdb    tRNAdb_ID          tRNAdb_organism
##          &lt;logical&gt; &lt;character&gt;  &lt;character&gt;              &lt;character&gt;
##   [1]         TRUE         RNA tdbR00000012 Saccharomyces cerevisiae
##   [2]         TRUE         RNA tdbR00000083 Saccharomyces cerevisiae
##   [3]         TRUE         RNA tdbR00000084 Saccharomyces cerevisiae
##       tRNAdb_strain tRNAdb_taxonomyID tRNAdb_verified
##         &lt;character&gt;       &lt;character&gt;       &lt;logical&gt;
##   [1]                            4932            TRUE
##   [2]                            4932            TRUE
##   [3]                            4932            TRUE
##                                                            tRNAdb_reference
##                                                             &lt;CharacterList&gt;
##   [1]       J.R.PENSWICK, R.MARTIN, G.DIRHEIMER (1975) FEBS LETT. 50, 28-31
##   [2] P.E.NIELSEN, V.LEICK (1982) BIOCHEM. BIOPHYS. RES. COMM. 106, 179-185
##   [3] G.KEITH, G.DIRHEIMER (1987) BIOCHEM. BIOPHYS. RES. COMM. 142, 183-187
##           tRNAdb_pmid
##       &lt;CharacterList&gt;
##   [1]                
##   [2]                
##   [3]                
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="co"># Saving the information as gff3 file</span>
<span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html">export.gff3</a></span>(<span class="kw">gff</span>, con = <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempfile</a></span>())
</pre></div>
<p>Please have a look at the <code>tRNA</code> package for further analysis of the tRNA sequences.</p>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
</pre></div>
<pre><code>## R version 4.0.2 (2020-06-22)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.4 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
## LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
## 
## locale:
##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] rtracklayer_1.48.0   tRNAdbImport_1.7.1   tRNA_1.6.0          
##  [4] Structstrings_1.4.0  Modstrings_1.4.0     Biostrings_2.56.0   
##  [7] XVector_0.28.0       GenomicRanges_1.40.0 GenomeInfoDb_1.24.2 
## [10] IRanges_2.22.2       S4Vectors_0.26.1     BiocGenerics_0.34.0 
## [13] BiocStyle_2.16.0    
## 
## loaded via a namespace (and not attached):
##  [1] lattice_0.20-41              Rsamtools_2.4.0             
##  [3] assertive.properties_0.0-4   assertive.types_0.0-3       
##  [5] assertive.data.us_0.0-2      assertthat_0.2.1            
##  [7] rprojroot_1.3-2              digest_0.6.25               
##  [9] R6_2.4.1                     backports_1.1.8             
## [11] evaluate_0.14                assertive.code_0.0-3        
## [13] httr_1.4.1                   ggplot2_3.3.2               
## [15] pillar_1.4.6                 assertive.strings_0.0-3     
## [17] zlibbioc_1.34.0              rlang_0.4.7                 
## [19] curl_4.3                     Matrix_1.2-18               
## [21] assertive_0.3-5.1            assertive.data_0.0-3        
## [23] rmarkdown_2.3.2              pkgdown_1.5.1.9000          
## [25] desc_1.2.0                   BiocParallel_1.22.0         
## [27] stringr_1.4.0                RCurl_1.98-1.2              
## [29] munsell_0.5.0                DelayedArray_0.14.1         
## [31] compiler_4.0.2               xfun_0.15                   
## [33] pkgconfig_2.0.3              htmltools_0.5.0             
## [35] SummarizedExperiment_1.18.2  tibble_3.0.3                
## [37] GenomeInfoDbData_1.2.3       bookdown_0.20               
## [39] matrixStats_0.56.0           assertive.sets_0.0-3        
## [41] codetools_0.2-16             XML_3.99-0.4                
## [43] crayon_1.3.4                 GenomicAlignments_1.24.0    
## [45] bitops_1.0-6                 grid_4.0.2                  
## [47] assertive.base_0.0-7         gtable_0.3.0                
## [49] lifecycle_0.2.0              magrittr_1.5                
## [51] assertive.models_0.0-2       scales_1.1.1                
## [53] stringi_1.4.6                fs_1.4.2                    
## [55] assertive.matrices_0.0-2     xml2_1.3.2                  
## [57] assertive.reflection_0.0-4.1 ellipsis_0.3.1              
## [59] assertive.datetimes_0.0-2    vctrs_0.3.2                 
## [61] tools_4.0.2                  Biobase_2.48.0              
## [63] glue_1.4.1                   assertive.numbers_0.0-2     
## [65] yaml_2.2.1                   colorspace_1.4-1            
## [67] BiocManager_1.30.10          assertive.files_0.0-2       
## [69] assertive.data.uk_0.0-2      memoise_1.1.0               
## [71] knitr_1.29</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Juehling-2009">
<p>Jühling, Frank, Mario Mörl, Roland K. Hartmann, Mathias Sprinzl, Peter F. Stadler, and Joern Pütz. 2009. “TRNAdb 2009: Compilation of tRNA Sequences and tRNA Genes.” <em>Nucleic Acids Research</em> 37: D159–D162. <a href="https://doi.org/10.1093/nar/gkn772">https://doi.org/10.1093/nar/gkn772</a>.</p>
</div>
<div id="ref-Sprinzl-2005">
<p>Sprinzl, Mathias, and Konstantin S. Vassilenko. 2005. “Compilation of tRNA Sequences and Sequences of tRNA Genes.” <em>Nucleic Acids Research</em> 33: D139–D140. <a href="https://doi.org/10.1093/nar/gki012">https://doi.org/10.1093/nar/gki012</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Felix G.M. Ernst.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
